gdal/ogr/ogrsf_frmts/flatgeobuf/geometryreader.cpp

```diff
@@ -259,7 +259,7 @@ OGRMultiPolygon *GeometryReader::readMultiPolygon()
     auto mp = std::unique_ptr<OGRMultiPolygon>(new OGRMultiPolygon());
     for (uoffset_t i = 0; i < parts->size(); i++) {
         GeometryReader reader { parts->Get(i), GeometryType::Polygon, m_hasZ, m_hasM };
-        auto p = std::unique_ptr<OGRPolygon>(reader.read()->toPolygon());
+        auto p = std::unique_ptr<OGRPolygon>(dynamic_cast<OGRPolygon *>(reader.read()));
```

Comment:
In that particular instance, I don't think the dynamic_cast<> is needed, as the reader is instructed to instanciate a GeometryType::Polygon. But ... in other places, such construct will result in a leak if read.read() returned a non-NULL pointer that is not of the type you dynamic_cast<> into

Reply:
Thanks, I think I understand your explanation and have changed that instance back.

Reply:
> have changed that instance back

Just for clarity: the other instances below that do dynamic_cast<> should be modified so that the unique_ptr is on the OGRGeometry* returned by reader.read(), not on the casted pointer

Reply:
Hmm I don't understand why I'd have to do that (dynamic cast should be safe for null?) and if I do, I don't see unique_ptr can be useful anymore as it cannot be casted?

Reply:
dynamic_cast is safe for null. But my point was that imagine that reader.read() returns a OGRPoint* instance, and you do std::unique_ptr<OGRLineString*>(dynamic_cast<OGRLineString*>(reader.read())) . In that case the object returned by reader.read() will be "lost" (leaking).
I think that for such cases, the std::unique_ptr makes things more complicated than the approach I took initially in 930b3c5d7e772282ffa90786e3e6d4fd0014ad2f

Reply:
Mm yes I see, that's a bit sad though.

Reply:
hum we could probably still use  the unique_ptr approach with something like:
```
auto poGeom = std::unique_ptr<OGRGeometry*>(reader.read());
if( !dynamic_cast<OGRLineString*>(poGeom.get()) // replace here OGRLineString by relevant type to check
     error_out;
poContainer->addGeometryDirectly(poGeom.release());
```

Reply:
But that would require a second dynamic_cast for collection types that expect a subtype?

Reply:
Not necessarily something like below should do, although not that pretty...
```
        auto geom = std::unique_ptr<OGRGeometry*>(reader.read());
        auto surf = dynamic_cast<OGRSurface*>(geom.get());
        if( surf == nullptr )
        {
            return nullptr;
        }
        geom.release(); // a bit weird to release without passing it explicitly, but we need to attach surf and not geom
        ms->addGeometryDirectly(surf);
```

Reply:
Mmm funny. :) I had a thought that if the dynamic_cast<OGRSurface*>(geom.get()) passes the nullptr check then it should be ok to use ->toSurface() on the geom?

Reply:
> then it should be ok to use ->toSurface() on the geom?

yes, you could do that too: ms->addGeometryDirectly(geom.release()->toSurface())